src/vsphere_cpi/lib/cloud/vsphere/nsxt_provider.rb

```diff
@@ -120,40 +126,59 @@ def add_vm_to_nsgroups(vm, ns_groups)
       return if nsxt_nics(vm).empty?
 
       logger.info("Adding vm '#{vm.cid}' to NSGroups: #{ns_groups}")
-      nsgroups = retrieve_nsgroups(ns_groups)
-
-      lports = logical_ports(vm)
-      nsgroups.each do |nsgroup|
-        logger.info("Adding LogicalPorts: #{lports.map(&:id)} to NSGroup '#{nsgroup.id}'")
-        grouping_obj_svc.add_or_remove_ns_group_expression(
-          nsgroup.id,
-          *to_simple_expressions(lports),
-          'ADD_MEMBERS'
-        )
+      Bosh::Retryable.new(
+          tries: 5,
+          sleep: ->(try_count, retry_exception) { 2 },
+          on: [NSXTApiCallError]
+      ).retryer do |i|
+        nsgroups = retrieve_nsgroups(ns_groups)
+        lports = logical_ports(vm)
+        nsgroups.each do |nsgroup|
+          logger.info("Adding LogicalPorts: #{lports.map(&:id)} to NSGroup '#{nsgroup.id}'")
+          begin
+            grouping_obj_svc.add_or_remove_ns_group_expression(
+              nsgroup.id,
+              *to_simple_expressions(lports),
+              'ADD_MEMBERS'
+            )
+          rescue NSXT::ApiCallError => e
+            retry if e.code == 409 || e.code == 412 #Conflict or PreconditionFailed
```

Comment:
The retryer block won't kick in unless we raise this error what it retries on.
`raise NSXTSecGrpUpdateError if e.code == 409 || e.code == 412`

Reply:
We do not need to use ruby's inbuilt `retry`

Can you look into how to make this exponentially backed off? `sleep: ->(try_count, retry_exception) { 2 },`

Reply:
`retryable.rb` has `sleep` parameter set to default as an exponential backoff. So if we do not pass it explicitly in the `Bosh::Retryable.new()`, it will execute the default option function `expontential sleeper`